README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# friends <img src='man/figures/logo.png' align="right" height="139" />

<!-- badges: start -->

[![R-CMD-check](https://github.com/EmilHvitfeldt/friends/workflows/R-CMD-check/badge.svg)](https://github.com/EmilHvitfeldt/friends/actions)
[![CRAN
status](http://www.r-pkg.org/badges/version/friends)](https://CRAN.R-project.org/package=friends)
[![Downloads](http://cranlogs.r-pkg.org/badges/friends)](https://CRAN.R-project.org/package=friends)
[![Lifecycle:
stable](https://img.shields.io/badge/lifecycle-stable-brightgreen.svg)](https://www.tidyverse.org/lifecycle/#stable)
<!-- badges: end -->

The goal of friends to provide the complete script transcription of the
[Friends](https://en.wikipedia.org/wiki/Friends) sitcom. The data
originates from the [Character
Mining](https://github.com/emorynlp/character-mining) repository which
includes references to scientific explorations using this data. This
package simply provides the data in tibble format instead of json files.

## Installation

You can install the released version of friends from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("friends")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("EmilHvitfeldt/friends")
```

## Example

The friends package comes with a couple of datasets. The main one is the
`friends` dataset which is a tibble containing all the utterances in the
show

``` r
library(friends)

friends
#> # A tibble: 67,373 x 6
#>    text                                speaker    season episode scene utterance
#>    <chr>                               <chr>       <int>   <int> <int>     <int>
#>  1 There's nothing to tell! He's just… Monica Ge…      1       1     1         1
#>  2 C'mon, you're going out with the g… Joey Trib…      1       1     1         2
#>  3 All right Joey, be nice. So does h… Chandler …      1       1     1         3
#>  4 Wait, does he eat chalk?            Phoebe Bu…      1       1     1         4
#>  5 (They all stare, bemused.)          Scene Dir…      1       1     1         5
#>  6 Just, 'cause, I don't want her to … Phoebe Bu…      1       1     1         6
#>  7 Okay, everybody relax. This is not… Monica Ge…      1       1     1         7
#>  8 Sounds like a date to me.           Chandler …      1       1     1         8
#>  9 [Time Lapse]                        Scene Dir…      1       1     1         9
#> 10 Alright, so I'm back in high schoo… Chandler …      1       1     1        10
#> # … with 67,363 more rows
```

All the utterances are broken down by `season`, `episode`, `scene` and
`utterance` which allows for very detailed analysis. Please note that
the `speaker` will be denoted `"Scene Directions"` to show scene
directions, or otherwise non-spoken descriptions.

The original data includes emotion and character entity annotation for
some of the utterances. These annotations have been included in separate
tibbles. These can easily be joined back to the main dataset as needed.

``` r
friends_entities
#> # A tibble: 10,557 x 5
#>    season episode scene utterance entities 
#>     <int>   <int> <int>     <int> <list>   
#>  1      1       1     1         2 <chr [1]>
#>  2      1       1     1         3 <chr [2]>
#>  3      1       1     1         4 <chr [1]>
#>  4      1       1     1         8 <chr [1]>
#>  5      1       1     1        31 <chr [2]>
#>  6      1       1     1        33 <chr [1]>
#>  7      1       1     1        34 <chr [1]>
#>  8      1       1     1        38 <chr [1]>
#>  9      1       1     1        42 <chr [4]>
#> 10      1       1     1        44 <chr [1]>
#> # … with 10,547 more rows

friends_emotions
#> # A tibble: 12,606 x 5
#>    season episode scene utterance emotion
#>     <int>   <int> <int>     <int> <chr>  
#>  1      1       1     4         1 Mad    
#>  2      1       1     4         3 Neutral
#>  3      1       1     4         4 Joyful 
#>  4      1       1     4         5 Neutral
#>  5      1       1     4         6 Neutral
#>  6      1       1     4         7 Neutral
#>  7      1       1     4         8 Scared 
#>  8      1       1     4        10 Joyful 
#>  9      1       1     4        11 Joyful 
#> 10      1       1     4        12 Sad    
#> # … with 12,596 more rows
```

There is also a tibble containing episode specific information such as
`title`, `air_date` and `imdb_rating`

``` r
friends_info
#> # A tibble: 236 x 8
#>    season episode title     directed_by  written_by  air_date   us_views_millio…
#>     <int>   <int> <chr>     <chr>        <chr>       <date>                <dbl>
#>  1      1       1 The Pilot James Burro… David Cran… 1994-09-22             21.5
#>  2      1       2 The One … James Burro… David Cran… 1994-09-29             20.2
#>  3      1       3 The One … James Burro… Jeffrey As… 1994-10-06             19.5
#>  4      1       4 The One … James Burro… Alexa Junge 1994-10-13             19.7
#>  5      1       5 The One … Pamela Frym… Jeff Green… 1994-10-20             18.6
#>  6      1       6 The One … Arlene Sanf… Adam Chase… 1994-10-27             18.2
#>  7      1       7 The One … James Burro… Jeffrey As… 1994-11-03             23.5
#>  8      1       8 The One … James Burro… Marta Kauf… 1994-11-10             21.1
#>  9      1       9 The One … James Burro… Jeff Green… 1994-11-17             23.1
#> 10      1      10 The One … Peter Bonerz Adam Chase… 1994-12-15             19.9
#> # … with 226 more rows, and 1 more variable: imdb_rating <dbl>
```

## Code of Conduct

Please note that the friends project is released with a [Contributor
Code of
Conduct](https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.


NEWS.md
# friends (development version)

* Deleted duplicated scene. Thanks @hope-data-science (#4)

# friends 0.1.0

* Added a `NEWS.md` file to track changes to the package.


R/data.R
#' The transcript of Friends
#'
#' Each season consists of episodes, each episode is divided into scenes,
#' each scene comprises utterances. One utterance per row.
#' Emotion annotation is included when available.
#'
#' @source \url{https://github.com/emorynlp/character-mining}
#'
#' @format A [tibble][tibble::tibble-package] with `r nrow(friends)` rows and
#'   `r ncol(friends)` variables:
#' \code{text}, \code{speaker}, \code{season}, \code{episode},
#' \code{scene} and \code{utterance}.
"friends"

#' Character Entities for transcript of Friends
#'
#' This [tibble][tibble::tibble-package] contains the character entities for the
#' utterances where it is available.
#'
#' @source \url{https://github.com/emorynlp/character-mining}
#'
#' @format A [tibble][tibble::tibble-package] with `r nrow(friends_entities)`
#'   rows and `r ncol(friends_entities)` variables:
#' \code{season}, \code{episode}, \code{scene}, \code{utterance} and
#' \code{entities}.
"friends_entities"

#' Emotions for transcript of Friends
#'
#' This [tibble][tibble::tibble-package] contains the emotions for the
#' utterances where it is available.
#'
#' @source \url{https://github.com/emorynlp/character-mining}
#' @source \url{https://github.com/emorynlp/emotion-detection}
#'
#' @format A [tibble][tibble::tibble-package] with `r nrow(friends_emotions)`
#'   rows and `r ncol(friends_emotions)` variables:
#' \code{season}, \code{episode}, \code{scene}, \code{utterance} and
#' \code{entities}.
"friends_emotions"

#' Episode Information
#'
#' This [tibble][tibble::tibble-package] contains additional information about
#' each of the episodes. Information was sourced from Wikipedia and IMDb on
#' August 26th, 2020.
#'
#' @source \url{https://en.wikipedia.org/wiki/Friends}
#' @source \url{https://www.imdb.com/title/tt0108778/}
#'
#' @format A [tibble][tibble::tibble-package] with `r nrow(friends_info)` rows
#'   and `r ncol(friends_info)` variables:
#' \code{season}, \code{episode}, \code{title}, \code{directed_by},
#' \code{written_by}, \code{air_date}, \code{us_views_millions} and
#' \code{imdb_rating}.
"friends_info"


data-raw/friends_info.R
## code to prepare `friends_wikipedia` dataset goes here

library(tidyverse)
library(rvest)
library(glue)

wikitable_raw <- read_html("https://en.wikipedia.org/wiki/List_of_Friends_episodes") %>%
  html_nodes('table[class="wikitable plainrowheaders wikiepisodetable"]') %>%
  html_table() %>%
  map2(seq_len(10),
       ~select(.x,
               episode = `No. inseason`,
               title = Title,
               directed_by = `Directed by`,
               written_by = `Written by`,
               air_date = `Original air date`,
               us_views_millions = `U.S. viewers(millions)`) %>%
         mutate(season = .y)) %>%
  bind_rows()

split_episode <- function(x) {
  x <- as.character(x)
  if (nchar(x) == 4) {
    return(paste(substr(x, 1, 2), substr(x, 3, 4)))
  }
  x
}

imdb_rating <- function(x) {
  rating <- read_html(glue("https://www.imdb.com/title/tt0108778/episodes?season={x}")) %>%
    html_nodes('div[class="list detail eplist"]
             div[class="ipl-rating-star small"]
             span[class="ipl-rating-star__rating"]') %>%
    html_text() %>%
    as.numeric()
  tibble(season = x,
         episode = seq_along(rating),
         imdb_rating = rating)
}

imdb_ratings <- map_dfr(seq_len(10), imdb_rating)

imdb_ratings <- imdb_ratings %>%
  add_row(season = 10,
          episode = 18,
          imdb_rating = tail(imdb_ratings$imdb_rating, 1))

friends_info <- wikitable_raw %>%
  mutate(episode = map_chr(episode, split_episode)) %>%
  separate_rows(episode, sep = " ", convert = TRUE) %>%
  mutate(title = str_remove(title, '"'),
         title = str_remove(title, '".*'),
         air_date = str_remove(air_date, "\\(.*"),
         air_date = as.Date(air_date, format = "%B %e, %Y"),
         us_views_millions = str_remove(us_views_millions, "\\[.*"),
         us_views_millions = as.numeric(us_views_millions)) %>%
  select(season, everything()) %>%
  left_join(imdb_ratings, by = c("season", "episode")) %>%
  mutate(across(season:episode, as.integer))

usethis::use_data(friends_info, overwrite = TRUE)



data-raw/friends_emotions.R
## code to prepare `friends_emotions` dataset goes here
library(tidyverse)
library(jsonlite)

extract_emotion <- function(x) {
  if (length(x) == 0) {
    return(NA)
  } else {
    x[[1]]
  }
}

read_utterances <- function(x) {
  res <- list(id = x$utterance_id,
              emotion = extract_emotion(x$emotion))
  res
}

read_scene <- function(x) {
  map_dfr(x$utterances, read_utterances)
}

read_episode <- function(x) {
  map_dfr(x$scenes, read_scene)
}

read_season <- function(x) {
  map_dfr(x$episodes, read_episode)
}

read_friends <- function() {
  season_numbers <- c("01", "02", "03", "04", "05", "06", "07", "08", "09", "10")
  urls <- glue::glue("https://raw.githubusercontent.com/emorynlp/character-mining/master/json/friends_season_{season_numbers}.json")
  map_dfr(urls, ~ .x %>% read_json() %>% read_season())
}

friends_raw <- read_friends()

friends_emotions <- friends_raw %>%
  filter(!is.na(emotion)) %>%
  separate(id, c("season", "episode", "scene", "utterance")) %>%
  mutate(across(c("season", "episode", "scene", "utterance"), parse_number)) %>%
  mutate(across(season:utterance, as.integer))

usethis::use_data(friends_emotions, overwrite = TRUE)


data-raw/friends_entities.R
## code to prepare `friends_entities` dataset goes here
library(tidyverse)
library(jsonlite)

extract_entities <- function(x) {
  map_chr(x[lengths(x) == 1],
          ~ .x[[1]][[3]])
}

read_utterances <- function(x) {
  res <- list(id = x$utterance_id,
              entities = list(extract_entities(x$character_entities)))
  res
}

read_scene <- function(x) {
  map_dfr(x$utterances, read_utterances)
}

read_episode <- function(x) {
  map_dfr(x$scenes, read_scene)
}

read_season <- function(x) {
  map_dfr(x$episodes, read_episode)
}

read_friends <- function() {
  season_numbers <- c("01", "02", "03", "04", "05", "06", "07", "08", "09", "10")
  urls <- glue::glue("https://raw.githubusercontent.com/emorynlp/character-mining/master/json/friends_season_{season_numbers}.json")
  map_dfr(urls, ~ .x %>% read_json() %>% read_season())
}

friends_raw <- read_friends()

friends_entities <- friends_raw %>%
  filter(0 < lengths(entities)) %>%
  separate(id, c("season", "episode", "scene", "utterance")) %>%
  mutate(across(c("season", "episode", "scene", "utterance"), parse_number)) %>%
  mutate(across(season:utterance, as.integer))

usethis::use_data(friends_entities, overwrite = TRUE)


data-raw/friends.R
library(tidyverse)
library(jsonlite)

extract_entities <- function(x) {
  map_chr(x[lengths(x) == 1],
          ~ .x[[1]][[3]])
}

read_utterances <- function(x) {
  res <- list(id = x$utterance_id,
              text = x$transcript)

  if (length(x$speakers) == 1) {
    res$speaker <- x$speakers[[1]]
  } else {
    res$speaker <- NA
  }

  if (!is.null(x$transcript_with_note)) {
    res$transcript_with_note <- x$transcript_with_note
  } else {
    res$transcript_with_note <- NA
  }

  res
}

read_scene <- function(x) {
  map_dfr(x$utterances, read_utterances)
}

read_episode <- function(x) {
  map_dfr(x$scenes, read_scene)
}

read_season <- function(x) {
  map_dfr(x$episodes, read_episode)
}

read_friends <- function() {
  season_numbers <- c("01", "02", "03", "04", "05", "06", "07", "08", "09", "10")
  urls <- glue::glue("https://raw.githubusercontent.com/emorynlp/character-mining/master/json/friends_season_{season_numbers}.json")
  map_dfr(urls, ~ .x %>% read_json() %>% read_season())
}

friends_raw <- read_friends()

friends <- friends_raw %>%
  mutate(speaker = if_else(nchar(text) == 0, "Scene Directions", speaker),
         text = if_else(nchar(text) == 0, transcript_with_note, text)) %>%
  separate(id, c("season", "episode", "scene", "utterance")) %>%
  mutate(across(c("season", "episode", "scene", "utterance"), parse_number)) %>%
  mutate(across(c("season", "episode", "scene", "utterance"), as.integer)) %>%
  select(-transcript_with_note) %>%
  select(text, speaker, season, episode, scene, utterance) %>%
  filter(!(season == 4 & episode == 21 & scene == 16)) %>%
  mutate(scene = if_else(season == 4 & episode == 21 & scene > 16, scene - 1L, scene))

usethis::use_data(friends, overwrite = TRUE)
